packages/contracts/test/hashchain.test.ts

```diff
@@ -0,0 +1,365 @@
+                    posRep.toString(),
+                    negRep.toString(),
+                    graffiti.toString(),
+                    timestamp.toString(),
+                ]
+            }
+        ),
+        dummyBalances,
+    ].flat()
+    const oldEpochKeysLeaves = [
+        hashchain.epochKeys.map((epk) => tree.getLeaf(epk.toBigInt())),
+        dummyBalances.map(() => defaultEpochTreeLeaf),
+    ].flat()
+    const circuitInputs = {
+        start_root: startRoot,
+        epoch_keys: allEpochKeys.map((k) => k.toString()),
+        epoch_key_balances: allBalances,
+        old_epoch_key_hashes: oldEpochKeysLeaves,
+        path_elements: allEpochKeys.map((key, i) => {
+            const p = tree.createProof(BigInt(key))
+            if (i < hashchain.epochKeys.length) {
+                const { posRep, negRep, graffiti, timestamp } =
+                    hashchain.epochKeyBalances[i]
+                tree.update(
+                    BigInt(key),
+                    hash4([posRep, negRep, graffiti, timestamp])
+                )
+            }
+            return p
+        }),
+        epoch: epoch.toString(),
+        attester_id: attester.address,
+        hashchain_index: hashchainIndex.toString(),
+        epoch_key_count: hashchain.epochKeys.length, // process epoch keys with attestations
+    }
+
+    return stringifyBigInts(circuitInputs)
+}
+
+describe('Hashchain tests', function () {
+    this.timeout(300000)
+    let unirepContract
+    let attester
+
+    before(async () => {
+        const accounts = await ethers.getSigners()
+        unirepContract = await deployUnirep(accounts[0])
+
+        attester = accounts[1]
+        await unirepContract
+            .connect(attester)
+            .attesterSignUp(EPOCH_LENGTH)
+            .then((t) => t.wait())
+    })
+
+    {
+        let snapshot
+        beforeEach(async () => {
+            snapshot = await ethers.provider.send('evm_snapshot', [])
+        })
+        afterEach(() => ethers.provider.send('evm_revert', [snapshot]))
+    }
+
+    it('should build a hashchain', async () => {
+        const epochKey = '0x12345'
+        const epoch = await unirepContract.attesterCurrentEpoch(
+            attester.address
+        )
+
+        await unirepContract
+            .connect(attester)
+            .submitAttestation(epoch, epochKey, 1, 0, 0)
+            .then((t) => t.wait())
+        await expect(unirepContract.buildHashchain(attester.address, epoch))
+            .to.emit(unirepContract, 'HashchainBuilt')
+            .withArgs(epoch, attester.address, 0)
+    })
+
+    it('should build and process a hashchain', async () => {
+        const epochKey = '0x123455'
+        const epoch = await unirepContract.attesterCurrentEpoch(
+            attester.address
+        )
+
+        await unirepContract
+            .connect(attester)
+            .submitAttestation(epoch, epochKey, 1, 0, 0)
+            .then((t) => t.wait())
+        const hashchainIndex = 0
```

Comment:
I think one of the problems is that how attester can get the current `hashchainIndex` that is going to be processed
maybe it can be written in the test
(in current contracts tests and core tests we just use `hashchainIndex` that we assign but not from contract status or events)
or in the `Synchronizer` it can handle `HashchainBuilt` and `HashchainProcessed` events so that we don't have to use `hashchainIndex` as input in `genAggregateEpochKeysProof`
https://github.com/Unirep/Unirep/blob/4b12237a1fb8bec43ca6ff57a224b17617238cec/packages/core/src/Synchronizer.ts#L389

Reply:
Right now they have to manually call `attesterHashchainIndex` on the unirep contract. Adding better handlers can certainly be done, but I'd prefer to do it in a separate pull request.

Reply:
an issue opened here: #249 and #248